sql/identities/identities.go

```diff
@@ -123,3 +129,62 @@ func GetMalicious(db sql.Executor) (nids []types.NodeID, err error) {
 	}
 	return nids, nil
 }
+
+// Married checks if id is married.
+// ID is married if it has non-null marriage_atx column.
+func Married(db sql.Executor, id types.NodeID) (bool, error) {
+	rows, err := db.Exec("select 1 from identities where pubkey = ?1 and marriage_atx is not null;",
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, id.Bytes())
+		}, nil)
+	if err != nil {
+		return false, fmt.Errorf("married %v: %w", id, err)
+	}
+	return rows > 0, nil
+}
+
+// Set marriage inserts marriage ATX for given identity.
+// If identitty doesn't exist - create it.
+func SetMarriage(db sql.Executor, id types.NodeID, atx types.ATXID) error {
+	_, err := db.Exec(`
+	INSERT INTO identities (pubkey, marriage_atx)
+	values (?1, ?2)
+	ON CONFLICT(pubkey) DO UPDATE SET marriage_atx = excluded.marriage_atx
+	WHERE marriage_atx IS NULL;`,
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, id.Bytes())
+			stmt.BindBytes(2, atx.Bytes())
+		}, nil,
+	)
+	if err != nil {
+		return fmt.Errorf("setting marriage %v: %w", id, err)
+	}
+	return nil
+}
+
+// EquivocationSet returns all node IDs that are married to the given node ID
+// including itself.
+func EquivocationSet(db sql.Executor, id types.NodeID) ([]types.NodeID, error) {
+	var ids []types.NodeID
+
+	rows, err := db.Exec(`
+	SELECT pubkey FROM identities
+	WHERE marriage_atx = (SELECT marriage_atx FROM identities WHERE pubkey = ?1) AND marriage_atx IS NOT NULL;`,
```

Comment:
The second half seems redundant?

```suggestion
	WHERE marriage_atx = (SELECT marriage_atx FROM identities WHERE pubkey = ?1);`,
```

Reply:
This is required to handle a case when the ID is not married (marriage_atx = NULL) to avoid returning all non-married IDs from the DB. In this case `rows == 0` and the function returns a single ID (non-married IDs form a set with just themselves).

Reply:
Theoretically an identity could publish an ATX only containing a certificate signed by itself creating an equivocation set that only contains themselves. The "self marry" certificate is unfortunately required, but we could syntactically check that the `MarriageCertificates` a) contains at least 2 entries and b) only contains distinct entries.